Automattic/woocommerce-payments
# Prevent onboarding flow access without server connection

Fixes #7360 

#### Changes proposed in this Pull Request
Redirect back to the connection page with an error message when attempting to access the onboarding flow without a server connection. Fixing the attached issue, which was caused by Jetpack redirecting to the onboarding flow after abandoning the auth process.

I've tried different approaches:
- My initial thought was that Jetpack should only redirect to the given URL after a successful auth, but after digging quite a bit on it, there is no straightforward way to redirect to the previous URL when the auth process is abandoned.
- Not registering the onboarding page when the server is not connected sound like a simple way to fix the issue, but the result is a "Sorry, you are not allowed to access this page." message, which doesn't feel accurate for this case.

So I opted to use the one in this PR, redirecting back to the connect screen showing "Please connect to WordPress.com to start using WooPayments.".

#### Testing instructions
We can use a JN site to test the real flow or also follow the following local instructions:
- Update [`is_server_connected`](https://github.com/Automattic/woocommerce-payments/blob/cabe96f71a15dc18d7e9e83416627ec496a30404/includes/wc-payment-api/class-wc-payments-api-client.php#L181) to return false.
- Ensure you don't have an onboarded account or enable `Force the WCPay plugin to act as disconnected from the WCPay Server` in WCPay Dev. Check also `Enable Progressive Onboarding`.
- Go to the connect page.
- Click on `Finish setup`.
- On the Jetpack connection screen click on `Return to [your site name]`
- It should redirect back to the connect screen and display the mentioned error message.

#### Screenshot
<img width="696" alt="Screenshot 2023-10-05 at 11 47 22" src="https://github.com/Automattic/woocommerce-payments/assets/7670276/7bb7b5ac-63d2-449f-8d3c-bdfa83fc2d54">

-------------------


**Post merge**
